### PR TITLE
One click

### PR DESCRIPTION
**Pull request by nilatti**
@schuyler1d 
This goes with Asana task https://app.asana.com/0/153988187951298/933603975155761

### CODE REVIEW

File: actionkit/utils.py [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import base64
-import datetime
+import Crypto.Random
+from datetime import datetime
 import hashlib
+import psycopg2
```
Comment (schuyler1d) [inline actionkit/utils.py]: This is a library file (and is imported in many projects for the methods here)
So: please remove the `psycopg2` and `settings` imports here.  `settings` can be passed as an argument to the methods here -- or even better, just pass the variables necessary into the abstract methods in this file.

File: actionkit/utils.py [hunk 2]
```diff
@@ -73,3 +76,51 @@ def date_in_range(akiddatestr):
...
+
+def payment_hash(token_id, datetime_format, braintree_secret, hash_separator):
+    parts = [
+        token_id,
+        datetime.now().strftime(datetime_format),
+        randstr32()[:8],
+    ]
+    return _append_hash(hash_separator, braintree_secret, *parts)
+
+def generate_token_id(user_id):
```
Comment (schuyler1d) [inline actionkit/utils.py]: Let's remove this method from the library and move this to a separate repo for the 'application' -- probably something we should understand at an Architecture level before committing to e.g. doing a SQL call per-userid.  This is a good prototype demonstrating feasibility, though.

File: actionkit/utils.py [hunk 3]
```diff
@@ -73,3 +76,51 @@ def date_in_range(akiddatestr):
...
+    con.close()
+    return data[0]
+
+def quickpay_url(datetime_format, user_id, hash_separator):
+    ak_secret = settings.AK_SECRET
+    braintree_secret = settings.BRAINTREE_ONECLICK_SECRET
+    akid = generate_akid(ak_secret, user_id)
+    token_id = generate_token_id(user_id)
+
+    return 'https://act.moveon.org/donate/civ-donation-quickpay?payment_hash=' + payment_hash(token_id, datetime_format, braintree_secret, hash_separator) + '&akid=' + akid
```
Comment (schuyler1d) [inline actionkit/utils.py]: This library should not assume MoveOn urls, so `page_url` could be an argument or we could just remove the function here, and, again, wait for a planned architecture and application

File: actionkit/quickpay_url_generator.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    hash_separator = '.'
+    quickpay_urls = []
+    for user_id in user_ids:
+        user_id = '.' + user_id
+        url = utils.quickpay_url(datetime_format, user_id, hash_separator)
+        print(url)
+        quickpay_urls.append(url)
+    return quickpay_urls
+
+print(get_array_of_payment_urls(['18715613', '8934025', '7666955']))
```
Comment (schuyler1d) [inline actionkit/quickpay_url_generator.py:32]: Better to do the `if __name__ == '__main__'` thing here, too -- or even just remove

File: actionkit/test_quickpay_url_generator.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+from datetime import datetime
+import re
+import settings
+import unittest
+import utils
+
+class PaymentUrlGeneratorTest(unittest.TestCase):
+    def setUp(self):
+        self.token_id = '3nmqoul2'
```
Comment (schuyler1d) [inline actionkit/test_quickpay_url_generator.py:10]: Just making sure this isn't a real token secret?

File: actionkit/test_quickpay_url_generator.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        self.user_id = '.' + '18715613'
+        self.datetime_format = '%Y%m%d%H%M'
+        self.hash_separator = '.'
+        self.braintree_secret = settings.BRAINTREE_ONECLICK_SECRET
+        self.ak_secret = settings.AK_SECRET
+    def test_quickpay_url(self): #all of these values depend on that user being the specific one it is.
+        url = utils.quickpay_url(self.datetime_format, self.user_id, self.hash_separator)
+        self.assertIn("act.moveon.org/donate/civ-donation-quickpay?", url)
+        self.assertIn("payment_hash=3nmqoul2", url)
+        self.assertIn("&akid=.18715613", url)
```
Comment (schuyler1d) [inline actionkit/test_quickpay_url_generator.py:30]: would be good to add a test that it matches `.18715613.\w{6}`

File: actionkit/test_quickpay_url_generator.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import unittest
+import utils
+
+class PaymentUrlGeneratorTest(unittest.TestCase):
+    def setUp(self):
+        self.token_id = '3nmqoul2'
+        self.user_id = '.' + '18715613'
+        self.datetime_format = '%Y%m%d%H%M'
+        self.hash_separator = '.'
+        self.braintree_secret = settings.BRAINTREE_ONECLICK_SECRET
```
Comment (schuyler1d) [inline actionkit/test_quickpay_url_generator.py:14]: This is a test, so let's set the values to junk plausible ones and then verify the actual hashes generated to ensure the algorithm 'works' as expected.

**Review by schuyler1d — CHANGES_REQUESTED** [review]
Hey sorry, forgot to hit submit here.

Basically: functions good, DB and moveon-specific code let's move elsewhere after a plan

File: actionkit/test_quickpay_url_generator.py (hunk 5 above)
Comment (nilatti) [reply actionkit/test_quickpay_url_generator.py:10]: It's mine--I didn't know how to generate a fake one. I felt like it was okay for this proof of concept phase, but not to go beyond that.

File: actionkit/utils.py (hunk 2 above)
Comment (nilatti) [reply actionkit/utils.py]: I agree with this and several of your other comments that are similar. I wasn't exactly sure how the architecture would work ultimately. This whole thing is a very prototypey prototype. Let's connect about what the next phase should look like, architecturally.

File: actionkit/test_quickpay_url_generator.py (hunk 7 above)
Comment (nilatti) [reply actionkit/test_quickpay_url_generator.py:14]: Do you have documentation about what would make them plausible enough to pass?

**Pull request closed without merging**